File: Examples/intel_examples/examples_core_c/vmlc/source/vzatan.c
/*******************************************************************************
* Copyright 2001-2020 Intel Corporation.
*
* This software and the related documents are Intel copyrighted  materials,  and
* your use of  them is  governed by the  express license  under which  they were
* provided to you (License).  Unless the License provides otherwise, you may not
* use, modify, copy, publish, distribute,  disclose or transmit this software or
* the related documents without Intel's prior written permission.
*
* This software and the related documents  are provided as  is,  with no express
* or implied  warranties,  other  than those  that are  expressly stated  in the
* License.
*******************************************************************************/

/*
!  Content:
!    vzAtan  Example Program Text
!******************************************************************************/

#include <stdio.h>
#include "mkl_vml.h"

#include "_rms.h"

#define INCA 3
#define INCB 5

int main()
{
  MKL_Complex16 zA[10],zB[10];
  MKL_Complex16 zBha0[10],zBha1[10],zBha2[10];
  MKL_Complex16           zBla1[10],zBla2[10];
  MKL_Complex16           zBep1[10],zBep2[10];
  float CurRMS,MaxRMS=0.0;

  MKL_Complex16 zA_I[10*INCA],zB_I[10*INCB];
  MKL_Complex16 zBha0_I[10*INCB],zBha1_I[10*INCB],zBha2_I[10*INCB];
  MKL_Complex16                  zBla1_I[10*INCB],zBla2_I[10*INCB];
  MKL_Complex16                  zBep1_I[10*INCB],zBep2_I[10*INCB];
  float CurRMS_I,MaxRMS_I=0.0;
  MKL_INT inca=INCA,incb=INCB;

  MKL_INT i=0,vec_len=10;

  zA[0].real=-10000.0000;zA[0].imag=10000.0000;
  zA[1].real=-7777.7777;zA[1].imag=7777.7777;
  zA[2].real=-5555.5555;zA[2].imag=5555.5555;
  zA[3].real=-3333.3333;zA[3].imag=3333.3333;
  zA[4].real=-1111.1111;zA[4].imag=1111.1111;
  zA[5].real=1111.1111;zA[5].imag=-1111.1111;
  zA[6].real=3333.3333;zA[6].imag=-3333.3333;
  zA[7].real=5555.5555;zA[7].imag=-5555.5555;
  zA[8].real=7777.7777;zA[8].imag=-7777.7777;
  zA[9].real=10000.0000;zA[9].imag=-10000.0000;
  zB[0].real=-1.5707463267948132e+000;zB[0].imag=4.9999999916666665e-005;
  zB[1].real=-1.5707320410797909e+000;zB[1].imag=6.4285714751457726e-005;
  zB[2].real=-1.5707063267935106e+000;zB[2].imag=9.0000000413999990e-005;
  zB[3].real=-1.5706463267911466e+000;zB[3].imag=1.4999999924999991e-004;
  zB[4].real=-1.5703463267296467e+000;zB[4].imag=4.4999994374998341e-004;
  zB[5].real=1.5703463267296467e+000;zB[5].imag=-4.4999994374998341e-004;
  zB[6].real=1.5706463267911466e+000;zB[6].imag=-1.4999999924999991e-004;
  zB[7].real=1.5707063267935106e+000;zB[7].imag=-9.0000000413999990e-005;
  zB[8].real=1.5707320410797909e+000;zB[8].imag=-6.4285714751457726e-005;
  zB[9].real=1.5707463267948132e+000;zB[9].imag=-4.9999999916666665e-005;

  for(i=0;i<10;i++) {
    zA_I[i*inca]=zA[i];
    zB_I[i*incb]=zB[i];
  }

  vzAtan(vec_len,zA,zBha0);
  vzAtanI(vec_len,zA_I,inca,zBha0_I,incb);

  vmzAtan(vec_len,zA,zBep1,VML_EP);
  vmzAtanI(vec_len,zA_I,inca,zBep1_I,incb,VML_EP);

  vmlSetMode(VML_EP);
  vzAtan(vec_len,zA,zBep2);
  vzAtanI(vec_len,zA_I,inca,zBep2_I,incb);

  vmzAtan(vec_len,zA,zBla1,VML_LA);
  vmzAtanI(vec_len,zA_I,inca,zBla1_I,incb,VML_LA);

  vmlSetMode(VML_LA);
  vzAtan(vec_len,zA,zBla2);
  vzAtanI(vec_len,zA_I,inca,zBla2_I,incb);

  vmzAtan(vec_len,zA,zBha1,VML_HA);
  vmzAtanI(vec_len,zA_I,inca,zBha1_I,incb,VML_HA);

  vmlSetMode(VML_HA);
  vzAtan(vec_len,zA,zBha2);
  vzAtanI(vec_len,zA_I,inca,zBha2_I,incb);

  for(i=0;i<10;i++) {
    if(zBha0[i].real!=zBha1[i].real || zBha0[i].imag!=zBha1[i].imag || zBha1[i].real!=zBha2[i].real ||
        zBha1[i].imag!=zBha2[i].imag) {
      printf("Error! Difference between vzAtan and vmzAtan in VML_HA mode detected.\n");
      return 1;
    }

    if(zBha0_I[i*incb].real!=zBha1_I[i*incb].real || zBha0_I[i*incb].imag!=zBha1_I[i*incb].imag ||
        zBha1_I[i*incb].real!=zBha2_I[i*incb].real || zBha1_I[i*incb].imag!=zBha2_I[i*incb].imag) {
      printf("Error! Difference between vzAtanI and vmzAtanI in VML_HA mode detected.\n");
      return 1;
    }

    if(zBla1[i].real!=zBla2[i].real || zBla1[i].imag!=zBla2[i].imag) {
      printf("Error! Difference between vzAtan and vmzAtan in VML_LA mode detected.\n");
      return 1;
    }

    if(zBla1_I[i*incb].real!=zBla2_I[i*incb].real || zBla1_I[i*incb].imag!=zBla2_I[i*incb].imag) {
      printf("Error! Difference between vzAtanI and vmzAtanI in VML_LA mode detected.\n");
      return 1;
    }

    if(zBep1[i].real!=zBep2[i].real || zBep1[i].imag!=zBep2[i].imag) {
      printf("Error! Difference between vzAtan and vmzAtan in VML_EP mode detected.\n");
      return 1;
    }

    if(zBep1_I[i*incb].real!=zBep2_I[i*incb].real || zBep1_I[i*incb].imag!=zBep2_I[i*incb].imag) {
      printf("Error! Difference between vzAtanI and vmzAtanI in VML_EP mode detected.\n");
      return 1;
    }
  }

  printf("vzAtan test/example program\n\n");
  printf("           Argument                           vzAtan\n");
  printf("===============================================================================\n");
  for(i=0;i<10;i++) {
    printf("   % .4f %+.4f*i      % .10f %+.10f*i\n",zA[i].real,zA[i].imag,zBha0[i].real,zBha0[i].imag);
    CurRMS=zrelerr(zB[i],zBha0[i]);
    if(CurRMS>MaxRMS) MaxRMS=CurRMS;
  }
  printf("\n");
  printf("vzAtanI test/example program\n\n");
  printf("           Argument                           vzAtan\n");
  printf("===============================================================================\n");
  for(i=0;i<10;i++) {
    printf("   % .4f %+.4f*i      % .10f %+.10f*i\n",zA_I[i*inca].real,zA_I[i*inca].imag,zBha0_I[i*incb].real,
        zBha0_I[i*incb].imag);
    CurRMS_I=zrelerr(zB_I[i*incb],zBha0_I[i*incb]);
    if(CurRMS_I>MaxRMS_I) MaxRMS_I=CurRMS_I;
  }
  printf("\n");
  if(MaxRMS>=1e-14) {
    printf("Error! Relative accuracy is %.16f\n",MaxRMS);
    return 1;
  }
  else {
    printf("Relative accuracy is %.16f\n",MaxRMS);
  }

  printf("\n");

  if(MaxRMS_I>=1e-14) {
    printf("Error! Relative strided accuracy is %.16f\n",MaxRMS_I);
    return 1;
  }
  else {
    printf("Relative strided accuracy is %.16f\n",MaxRMS_I);
  }
  return 0;
}

File: Examples/intel_examples/examples_core_f/vmlf/source/vzatan.f
!===============================================================================
! Copyright 2001-2020 Intel Corporation.
!
! This software and the related documents are Intel copyrighted  materials,  and
! your use of  them is  governed by the  express license  under which  they were
! provided to you (License).  Unless the License provides otherwise, you may not
! use, modify, copy, publish, distribute,  disclose or transmit this software or
! the related documents without Intel's prior written permission.
!
! This software and the related documents  are provided as  is,  with no express
! or implied  warranties,  other  than those  that are  expressly stated  in the
! License.
!===============================================================================

!  Content:
!    vzAtan  Example Program Text
!*******************************************************************************

      include "_rms.fi"

      program MKL_VML_TEST

      include "mkl_vml.f90"
      real(kind=4) :: srelerr
      real(kind=8) :: drelerr
      real(kind=4) :: crelerr
      real(kind=8) :: zrelerr

      complex(kind=8) zA(10)
      complex(kind=8) zB(10)
      complex(kind=8) zBha0(10)
      complex(kind=8) zBha1(10)
      complex(kind=8) zBha2(10)
      complex(kind=8) zBla1(10)
      complex(kind=8) zBla2(10)
      complex(kind=8) zBep1(10)
      complex(kind=8) zBep2(10)

      real(kind=8) CurRMS,MaxRMS

      integer, parameter :: inca=3
      integer, parameter :: incb=5
      complex(kind=8) zA_I(10*inca)
      complex(kind=8) zB_I(10*incb)
      complex(kind=8) zBha0_I(10*incb)
      complex(kind=8) zBha1_I(10*incb)
      complex(kind=8) zBha2_I(10*incb)
      complex(kind=8) zBla1_I(10*incb)
      complex(kind=8) zBla2_I(10*incb)
      complex(kind=8) zBep1_I(10*incb)
      complex(kind=8) zBep2_I(10*incb)
      real(kind=8) CurRMS_I,MaxRMS_I

      integer(kind=8) mode
      integer tmode
      integer i, vec_len

      vec_len=10
      MaxRMS=0.0
      MaxRMS_I=0.0

      zA( 1)=(-10000.0000d+000,10000.0000d+000)
      zA( 2)=(-7777.7777d+000,7777.7777d+000)
      zA( 3)=(-5555.5555d+000,5555.5555d+000)
      zA( 4)=(-3333.3333d+000,3333.3333d+000)
      zA( 5)=(-1111.1111d+000,1111.1111d+000)
      zA( 6)=(1111.1111d+000,-1111.1111d+000)
      zA( 7)=(3333.3333d+000,-3333.3333d+000)
      zA( 8)=(5555.5555d+000,-5555.5555d+000)
      zA( 9)=(7777.7777d+000,-7777.7777d+000)
      zA(10)=(10000.0000d+000,-10000.0000d+000)
      zB( 1)=(-1.5707463267948132d+000,4.9999999916666665d-005)
      zB( 2)=(-1.5707320410797909d+000,6.4285714751457726d-005)
      zB( 3)=(-1.5707063267935106d+000,9.0000000413999990d-005)
      zB( 4)=(-1.5706463267911466d+000,1.4999999924999991d-004)
      zB( 5)=(-1.5703463267296467d+000,4.4999994374998341d-004)
      zB( 6)=(1.5703463267296467d+000,-4.4999994374998341d-004)
      zB( 7)=(1.5706463267911466d+000,-1.4999999924999991d-004)
      zB( 8)=(1.5707063267935106d+000,-9.0000000413999990d-005)
      zB( 9)=(1.5707320410797909d+000,-6.4285714751457726d-005)
      zB(10)=(1.5707463267948132d+000,-4.9999999916666665d-005)

      do i=1,10
          zA_I((i-1)*inca+1)=zA(i)
          zB_I((i-1)*incb+1)=zB(i)
      end do

      call VZATAN(vec_len,zA,zBha0)
      call VZATANI(vec_len,zA_I,inca,zBha0_I,incb)

      mode=VML_EP
      call VMZATAN(vec_len,zA,zBep1,mode)
      call VMZATANI(vec_len,zA_I,inca,zBep1_I,incb,mode)

      tmode=VML_EP
      tmode=VMLSETMODE(tmode)
      call VZATAN(vec_len,zA,zBep2)
      call VZATANI(vec_len,zA_I,inca,zBep2_I,incb)

      mode=VML_LA
      call VMZATAN(vec_len,zA,zBla1,mode)
      call VMZATANI(vec_len,zA_I,inca,zBla1_I,incb,mode)

      tmode=VML_LA
      tmode=VMLSETMODE(tmode)
      call VZATAN(vec_len,zA,zBla2)
      call VZATANI(vec_len,zA_I,inca,zBla2_I,incb)

      mode=VML_HA
      call VMZATAN(vec_len,zA,zBha1,mode)
      call VMZATANI(vec_len,zA_I,inca,zBha1_I,incb,mode)

      tmode=VML_HA
      tmode=VMLSETMODE(tmode)
      call VZATAN(vec_len,zA,zBha2)
      call VZATANI(vec_len,zA_I,inca,zBha2_I,incb)

      do i=1,10
          if(zBha0(i) .ne. zBha1(i)) then
              print *,"Error! Difference between VZATAN and"
              print *," VMZATAN in VML_HA mode detected"
              stop 1
          endif
          if(zBha0_I((i-1)*incb+1) .ne. zBha1_I((i-1)*incb+1)) then
              print *,"Error! Difference between VZATANI and"
              print *," VMZATANI in VML_HA mode detected"
              stop 1
          endif
          if(zBha1(i) .ne. zBha2(i)) then
              print *,"Error! Difference between VZATAN and"
              print *," VMZATAN in VML_HA mode detected"
              stop 1
          endif
          if(zBha1_I((i-1)*incb+1) .ne. zBha2_I((i-1)*incb+1)) then
              print *,"Error! Difference between VZATANI and"
              print *," VMZATANI in VML_HA mode detected"
              stop 1
          endif
          if(zBla1(i) .ne. zBla2(i)) then
              print *,"Error! Difference between VZATAN and"
              print *," VMZATAN in VML_LA mode detected"
              stop 1
          endif
          if(zBla1_I((i-1)*incb+1) .ne. zBla2_I((i-1)*incb+1)) then
              print *,"Error! Difference between VZATANI and"
              print *," VMZATANI in VML_LA mode detected"
              stop 1
          endif
          if(zBep1(i) .ne. zBep2(i)) then
              print *,"Error! Difference between VZATAN and"
              print *," VMZATAN in VML_EP mode detected"
              stop 1
          endif
          if(zBep1_I((i-1)*incb+1) .ne. zBep2_I((i-1)*incb+1)) then
              print *,"Error! Difference between VZATANI and"
              print *," VMZATANI in VML_EP mode detected"
              stop 1
          endif
      end do

      print *,"vzAtan test/example program"
      print *,""
      print *,"           Argument                     vzAtan"
      print *,"======================================================", &
     &        "========================"
      do i=1,vec_len
            print 10,zA(i),"    ",zBha0(i)
            CurRMS=zrelerr(zB(i),zBha0(i))
            if(CurRMS>MaxRMS) MaxRMS=CurRMS
      end do

      print *,""
      print *,"vzAtanI test/example program"
      print *,"           Argument                     vzAtan"
      print *,"======================================================", &
     &        "========================"
      do i=1,10
            print 10,zA_I((i-1)*inca+1),"    ",zBha0_I((i-1)*incb+1)
            CurRMS_I=zrelerr(zB_I((i-1)*incb+1),zBha0_I((i-1)*incb+1))
            if(CurRMS_I>MaxRMS_I) MaxRMS_I=CurRMS_I
      end do

      print *,""
      if(MaxRMS>=1e-14) then
            print 11,"Error! Relative accuracy is ",MaxRMS
            stop 1
      else
            print 11,"Relative accuracy is ",MaxRMS
      endif

      if(MaxRMS_I>=1e-14) then
            print 11,"Error! Relative strided accuracy is ",MaxRMS_I
            stop 1
      else
            print 11,"Relative strided accuracy is ",MaxRMS_I
      endif

10    format(E15.7,E15.7,A5,E15.7,E15.7)
11    format(A,F25.16)

      end
